### PR TITLE
Mesh SDF: pre-release API cleanup and correctness fixes

### PR DESCRIPTION
**Pull request by peterdsharpe**
# PhysicsNeMo Pull Request

## Description

Pre-release cleanup and hardening of the mesh-native SDF (`physicsnemo.mesh.spatial`), which landed after v2.1.1 and has not yet shipped in a release — so these API changes are still free. Four commits, independently revertable:

**1. API cleanup (`4e32f65`)**

- Rename `signed_distance_field_mesh` → `signed_distance_field`: the function lives in `physicsnemo.mesh.spatial`, so the `_mesh` suffix was redundant. (The array-based custom op `physicsnemo.nn.functional.signed_distance_field` is a different function with a different signature; the namespaces disambiguate, as in `torch.norm` / `torch.linalg.norm`.)
- Type `winding_backend` as `Literal["clustertree", "bruteforce"]` (`WindingBackend` type alias) instead of `str`, so type checkers validate call sites; runtime validation derives from the same `Literal` via `get_args` and now runs at function entry — before the BVH build and nearest-triangle search — instead of after them.
- Validate mesh/query device mismatch and negative `max_dist` up front with clear messages.
- Reflow a docstring line that ran far past the line width (`E501` is disabled repo-wide and `ruff format` does not reflow docstrings, which is how it slipped through).

**2. Degenerate-triangle correctness fix (`70c96fe`)**

Ericson's Voronoi-region closest-point classification assumes a non-degenerate triangle. On zero-area faces (repeated vertices or collinear points) several region tests fire vacuously and the cascade returns the wrong feature — always overestimating the distance. Measured against an exact union-of-edges oracle: ~30–70% wrong closest points for isolated repeated-vertex faces (distance errors up to ~7% relative), 1.5% for exactly-collinear faces, plus float32 misclassification of eps-scale slivers. The Triton kernel mirrors the same cascade and shared the bug. Degenerate junk faces coincident with valid geometry are masked by the distance min; isolated ones silently corrupt the SDF.

Fix: repair the geometry once per call instead of guarding the hot kernels — `_build_surface_mesh` displaces the off-edge vertex of any face with relative height < 1e-4 to the longest-edge midpoint plus a perpendicular offset `h = max(1e-4·L, 8·eps_f32·max|coord|)`. SDF and hit points move by at most `2h`, only for queries whose nearest face was degenerate; the pass is fixed-shape tensor code (sync-free) and leaves valid faces bit-identical. An exact per-pair in-kernel fallback was implemented first and rejected: it measured **20–25% slower kernel-only** (SIMT straight-line code pays for it on every candidate), whereas this approach keeps both closest-point kernels byte-identical to `main`.

**3. Hit-face indices via a named tuple (`d99c194`)**

`signed_distance_field` now returns `SignedDistanceFieldResult(sdf, hit_points, hit_faces)`. Both backends already computed the nearest-face index and threw it away; exposing it enables oriented face-normal lookups at the hit point (needed by commit 4). `hit_faces` is `-1` for queries beyond a finite `max_dist`. Doing this pre-release keeps it a clean signature change rather than a compatibility shim.

**4. Near-wall SDF normals fix in the unified aero recipe (`4336b05`)**

`ComputeSDFFromBoundary` fell back to the direction-away-from-surface-centroid for points within an absolute 1e-6 of the surface. That direction is unrelated to the wall normal — on a fender-like panel it is *tangent* to the surface (~80–90° median error vs the true normal) — and DrivAerML boundary layers put 1–3% of every 200k-point batch inside the band, so every volume model trained on this pipeline received corrupted normals concentrated exactly in the boundary layer. (The pattern traces to DoMINO, where direction-from-center-of-mass was a *separate* positional-encoding input, not a normals fallback.) The fix substitutes the oriented hit-face normal — exact at any wall distance including zero — inside a scale-aware band (`128·eps` per unit coordinate magnitude, since closest-point rounding noise grows with coordinate scale and an absolute cutoff under-covers geometry far from the origin).

**Verification**

- Existing SDF suite extended: 50+ tests pass on CPU and CUDA (Triton path included), plus new tests for the winding backends, degenerate-face repair (unit + end-to-end), hit-face indices (including the Triton sorted-order mapping), and the recipe's near-wall normals (closed elongated box mirroring the fender failure mode, wall distances down to 1e-8).
- Degenerate closest-point behavior fuzzed against an exact union-of-edges oracle (500k queries across repeated-vertex/collinear/collapsed patterns: 0 wrong) and a float32 sliver sweep (200k queries, heights 1e-3…1e-12: 0 over 1e-4 error).
- The `test_sdf_no_winding_path_is_sync_free` guard confirms the repair pass introduces no host synchronization.

**Known follow-up (out of scope here):** `physicsnemo.datapipes.transforms.geometric.ComputeNormals` has the same centroid-fallback pattern, but it is a raw-TensorDict transform with no mesh access, so the face-normal fix needs a data-contract change there.

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Dependencies

None.

### CODE REVIEW

**Comment by peterdsharpe** [comment]
/ok to test https://github.com/NVIDIA/physicsnemo/pull/1815/commits/4336b0570fc708a5bd96c7eed5dabdf22a517b8e

File: examples/cfd/external_aerodynamics/unified_external_aero_recipe/src/sdf.py [hunk 1]
```diff
@@ -135,17 +138,30 @@ def apply_to_domain(self, domain: DomainMesh) -> DomainMesh:
...
+            # pointing into the body like its neighbors (the SDF treats
+            # on-surface as outside, so dist == 0 gets the outward normal).
+            # The band is scale-aware: the closest point carries rounding
+            # noise ~ eps * |coordinate|, so an absolute cutoff under-covers
+            # geometry far from the origin. 128 eps (~1.5e-5 per unit
+            # coordinate) clears that noise floor with a wide margin, and
+            # widening the band is free because the substitute is exact.
+            # Computed unconditionally and selected with a mask rather than
+            # branching on ``near_surface.any()`` -- that host readback would
+            # stall the prefetch stream.
```
Comment (coreyjadams) [inline examples/cfd/external_aerodynamics/unified_external_aero_recipe/src/sdf.py:156]: Do we need so much verbosity here?

**Review by ktangsali — APPROVED** [review]

**Review by coreyjadams — APPROVED** [review]
This all looks fine.  I'm rubber stamping the _sdf_triton updates and related changes because I intend to purge those shortly, but good to go.

File: examples/cfd/external_aerodynamics/unified_external_aero_recipe/src/sdf.py (hunk 1 above)
Comment (peterdsharpe) [reply examples/cfd/external_aerodynamics/unified_external_aero_recipe/src/sdf.py:156]: I actually think every sentence here is useful information and non-obvious from the code itself (or at least, the *reasoning for why it is this way* is non-obvious), so I would vote to keep as-is in this instance, if that's ok!

**Comment by peterdsharpe** [comment]
/ok to test https://github.com/NVIDIA/physicsnemo/pull/1815/commits/4336b0570fc708a5bd96c7eed5dabdf22a517b8e

File: examples/cfd/external_aerodynamics/unified_external_aero_recipe/src/sdf.py [hunk 2]
```diff
@@ -135,17 +138,30 @@ def apply_to_domain(self, domain: DomainMesh) -> DomainMesh:
...
-            on_surface = dist < 1e-6
-            # The mesh stays intact; read its points only here, at point of use.
-            centroid = surface.points.float().mean(dim=0, keepdim=True)
-            normals = torch.where(
-                on_surface.unsqueeze(-1), query_points - centroid, normals
+            coord_scale = query_points.abs().amax(dim=-1).clamp(min=1.0)
+            near_surface = dist < (128.0 * torch.finfo(torch.float32).eps * coord_scale)
+            face_normals = surface.cell_normals.to(query_points.dtype)[hit_faces]
+            oriented = torch.where(
+                (sdf_values >= 0).unsqueeze(-1), face_normals, -face_normals
```
[the remark this answers is not in the document]
Comment (peterdsharpe) [reply examples/cfd/external_aerodynamics/unified_external_aero_recipe/src/sdf.py:166]: Fixed in 527396f7, with one correction to the analysis: `cell_normals` returns a **zero vector** for zero-area faces (the normal computation clamps the norm), not NaN — verified empirically. So the failure mode was a zero normal being substituted, not NaN propagation. The substitution is now guarded by a unit-norm check on the cell normal; degenerate hit faces keep the raw closest-point direction, and a regression test covers a degenerate face that is genuinely the nearest surface.

File: physicsnemo/mesh/spatial/sdf.py [hunk 3]
```diff
@@ -1032,9 +1217,11 @@ def signed_distance_field_mesh(
     hit_points = queries.clone()
 
     if n_queries == 0:
-        sdf = sdf.reshape(query_shape[:-1]).to(out_dtype)
-        hit_points = hit_points.reshape(query_shape).to(out_dtype)
-        return sdf, hit_points
+        return SignedDistanceFieldResult(
+            sdf=sdf.reshape(query_shape[:-1]).to(out_dtype),
+            hit_points=hit_points.reshape(query_shape).to(out_dtype),
+            hit_faces=torch.zeros(query_shape[:-1], dtype=torch.long, device=device),
```
[the remark this answers is not in the document]
Comment (peterdsharpe) [reply physicsnemo/mesh/spatial/sdf.py]: Done in 527396f7 — switched to `torch.full(..., -1)`. The tensor is always empty on that path, so this is purely contract clarity.

**Comment by peterdsharpe** [comment]
/ok to test https://github.com/NVIDIA/physicsnemo/pull/1815/commits/527396f7d20d566953c2cedd255c822ac18609b9
